scripts/demangle.py

```diff
@@ -97,50 +99,81 @@ def main(proj_name: str, proj_language: str = "c"):
             splited_file_func_name = file_func_name.split(FILE_FUNC_DELIM)
             file_path, mangle_func_name = (
                 splited_file_func_name[0],
                 splited_file_func_name[1],
             )
 
-            # if this is c file then the mangle does not exist
-            try:
-                demangle_func_name = cpp_demangle.demangle(mangle_func_name)
-            except ValueError:
-                demangle_func_name = mangle_func_name
-            if demangle_func_name == None:
-                warning(f" {mangle_func_name} demangle incorrect or unable to demangle")
-                data[cnt][
-                    file_func_name
-                ] = f" {mangle_func_name}demangle incorrect or unable to demangle"
-            else:
-                func_name = demangle_func_name.split("(")[0]
-                # Get code path from local
-                code_path = get_source_code_path(file_path, output_path)
-                if code_path == None:
-                    warning(f"Get source code path error {file_path}")
+            # Get code path from local
+            code_path = get_source_code_path(file_path, output_path)
+            if code_path == None:
+                warning(f"Get source code path error {file_path}")
+                data[cnt][file_func_name] = {
+                    "code": "Path Error",
+                    "data": data[cnt][file_func_name],
+                }
+                continue
+
+            # Get function content
+            temp_proj_language = proj_language
+            # Process using source code language
+            if proj_language == "python":
+                # Class_name need to defined in json file
+                # Have to update using CODE_EXTRACTOR for python if class_name needs to be defined
+                mangle_func_name = mangle_func_name.split("(")[0]
+            elif proj_language == "c++" or proj_language == "c":
+                # Get file extension for language of source code (since some C++ projects have some C source code)
+                filename_regex = ".+\.([A-Za-z]+)"
+                match = re.match(filename_regex, file_path)
+                proj_language = match.group(1)
+
+            func_content = CODE_EXTRACTOR[proj_language](code_path, mangle_func_name)
+            # Check func_content
+            if func_content == None:
+                if (
+                    proj_language == "cpp"
+                    or proj_language == "cxx"
+                    or proj_language == "cc"
+                ):
+                    try:
+                        # Check to see if template function, otherwise cannot find source code
+                        demangled_func_name = cpp_demangle.demangle(mangle_func_name)
+                        if "<" in demangled_func_name and ">" in demangled_func_name:
+                            data[cnt][file_func_name] = {
+                                "code": "Template Function",
```

Comment:
We should define some enum for status. For example
```py
from enum import Enum

class SourceCodeStatus(Enum):
    SUCCESS = 1
    TEMPLATE = 2
    NOT_FOUND = 3
```
and use this for another field "status", just leave the "code" field `None` if the status is not SUCCESS. In this way, if we need to filter on code status in the future, we don't have to do expensive string comparisons.

Reply:
Good idea.